lib/patch-https-request-for-proxying.js

```diff
@@ -0,0 +1,104 @@
+const {brokerServerUrl} = require('./config');
+const brokerServer = url.parse(brokerServerUrl || '');
+brokerServer.port = brokerServer.port ||
+  brokerServer.protocol === 'https:' ? '443' : '80';
+const httpsProxy = process.env.https_proxy || process.env.HTTPS_PROXY;
+
+// adapted from https://github.com/request/request/master/lib/getProxyFromURI.js
+
+function formatHostname(hostname) {
+  // canonicalize the hostname, so that 'oogle.com' won't match 'google.com'
+  return hostname.replace(/^\.*/, '.').toLowerCase();
+}
+
+function parseNoProxyZone(zone) {
+  zone = zone.trim().toLowerCase();
+
+  const zoneParts = zone.split(':', 2);
+  const zoneHost = formatHostname(zoneParts[0]);
+  const zonePort = zoneParts[1];
+  const hasPort = zone.indexOf(':') > -1;
+
+  return {hostname: zoneHost, port: zonePort, hasPort: hasPort};
+}
+
+function uriInNoProxy(uri, noProxy) {
+  const port = uri.port || (uri.protocol === 'https:' ? '443' : '80');
+  const hostname = formatHostname(uri.hostname);
+  const noProxyList = noProxy.split(',');
+
+  // iterate through the noProxyList until it finds a match.
+  return noProxyList.map(parseNoProxyZone).some(function (noProxyZone) {
+    const isMatchedAt = hostname.indexOf(noProxyZone.hostname);
+    const hostnameMatched = (
+      isMatchedAt > -1 &&
+        (isMatchedAt === hostname.length - noProxyZone.hostname.length));
+
+    if (noProxyZone.hasPort) {
+      return (port === noProxyZone.port) && hostnameMatched;
+    }
+
+    return hostnameMatched;
+  });
+}
+
+function shouldProxy(uri) {
+  // Decide the proper request proxy to use based on the request URI object and the
+  // environmental variables (NO_PROXY, HTTP_PROXY, etc.)
+  // respect NO_PROXY environment variables (see: http://lynx.isc.org/current/breakout/lynx_help/keystrokes/environments.html)
+
+  const noProxy = process.env.NO_PROXY || process.env.no_proxy || '';
+
+  // if the noProxy is a wildcard then return null
+
+  if (noProxy === '*') {
+    return false;
+  }
+
+  // if the noProxy is not empty and the uri is found return null
+
+  if (noProxy !== '' && uriInNoProxy(uri, noProxy)) {
+    return false;
+  }
+
+  // we should proxy
+  return true;
+}
+
+// Entry point: To patch or not to patch?
+if (httpsProxy && shouldProxy(brokerServer)) {
+  const proxy = url.parse(httpsProxy);
+
+  let Tunnel;
+  // Tunnel = tunnel.httpsOverHttps;
+  // if (proxy.protocol === 'http:' || (!proxy.protocol && proxy.port === '80')) {
+  Tunnel = tunnel.httpsOverHttp;
+  // }
+  const tunnelingAgent = Tunnel({
+    proxy: {
+      host: proxy.hostname,
+      port: proxy.port
+    }
+  });
+
+  // actual monkey patching: BEWARE!
+  const https = require('https');
+  const oldhttpsreq = https.request;
+  https.request = function (options, callback) {
+    if (options.host === brokerServer.host &&
+        // use == not === to match string and int due to options diffs
+        (options.port && options.port == brokerServer.port)) {
```

Comment:
This code means "don't proxy if `options.port` is missing" - do we want that? How about:
```
if (options.host === brokerServer.host) {
  if (options.port == brokerServer.port || (!options.port && brokerServer.port == 80)) {
    ...
```